File: nhitomi-idx/nhitomi/Discord/Commands/BookReadMessage.cs
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Discord;
using nhitomi.Database;
using nhitomi.Localization;

namespace nhitomi.Discord.Commands
{
    public class BookReadMessage : InteractiveMessage, IListJumpTriggerTarget
    {
        readonly ILocale _l;
        readonly ILinkGenerator _link;

        public (DbBook book, DbBookContent content) Book { get; set; }
        public int Position { get; private set; }
        public int End => Book.content.Pages.Length - 1;

        public BookReadMessage(nhitomiCommandContext context, ILinkGenerator link)
        {
            _l    = context.Locale.Sections["view.book"];
            _link = link;
        }

        protected override IEnumerable<ReactionTrigger> CreateTriggers()
        {
            foreach (var trigger in base.CreateTriggers())
                yield return trigger;

            yield return new ListTrigger(this, ListTriggerDirection.Left);
            yield return new ListTrigger(this, ListTriggerDirection.Right);
            yield return new ListJumpTrigger(this, ListJumpTriggerDestination.Input);
            yield return new DestroyTrigger(this);
        }

        public Task<bool> SetPositionAsync(int position, CancellationToken cancellationToken = default)
        {
            if (0 <= position && position <= End)
            {
                Position = position;
                return Task.FromResult(true);
            }

            return Task.FromResult(false);
        }

        protected override ReplyContent Render()
        {
            var (book, content) = Book;

            return new ReplyContent
            {
                Embed = new EmbedBuilder
                {
                    Title       = book.PrimaryName,
                    Description = _l["pagination", new { current = Position + 1, total = End + 1 }],
                    Url         = _link.GetWebLink($"books/{book.Id}/contents/{content.Id}?auth=discord"),
                    ImageUrl    = _link.GetApiLink($"books/{book.Id}/contents/{content.Id}/pages/{Position}"),
                    Color       = Color.DarkGreen,

                    Author = new EmbedAuthorBuilder
                    {
                        Name    = book.TagsArtist?.Length > 0 ? string.Join(", ", book.TagsArtist) : content.Source.ToString(),
                        IconUrl = _link.GetWebLink($"assets/icons/{content.Source}.jpg")
                    },

                    Footer = new EmbedFooterBuilder
                    {
                        Text = $"{book.Id}/{content.Id}"
                    }
                }
            };
        }
    }
}

File: nhitomi-discord/src/Messages/BookReadMessage.ts
import { Book, BookContent } from "nhitomi-api";
import { ListTrigger } from "../Triggers/ListTrigger";
import { DestroyTrigger } from "../Triggers/DestroyTrigger";
import { ReactionTrigger } from "../Interactive/trigger";
import { InteractiveMessage } from "../Interactive/message";
import { MessageContext } from "../context";
import { getApiLink, getWebLink } from "../client";
import { ListJumpTrigger } from "../Triggers/ListJumpTrigger";

export class BookReadMessage extends InteractiveMessage {
  constructor(context: MessageContext, readonly book: Book, readonly content: BookContent) {
    super(context);
  }

  position = 0;

  get length() {
    return this.content.pageCount;
  }

  protected async render() {
    this.position = Math.max(0, Math.min(this.length - 1, this.position));

    return {
      message: this.content.sourceUrl,
      embed: {
        title: this.book.primaryName,
        description: this.context.locale.get("read.pagination", {
          current: this.position + 1,
          total: this.length,
        }),
        url: getWebLink(`books/${this.book.id}/contents/${this.content.id}`),
        image: {
          url: getApiLink(`books/${this.book.id}/contents/${this.content.id}/pages/${this.position}`),
        },
        color: "DARK_GREEN",
        author: {
          name: (this.book.tags.artist || this.book.tags.circle || [this.content.source]).sort().join(", "),
          iconURL: getWebLink(`assets/icons/${this.content.source}.jpg`),
        },
        footer: {
          text: `${this.book.id}/${this.content.id}`,
        },
      },
    };
  }

  protected createTriggers(): ReactionTrigger[] {
    return [
      ...super.createTriggers(),

      new ListTrigger(this, "left"),
      new ListTrigger(this, "right"),
      new ListJumpTrigger(this, "input"),
      new DestroyTrigger(this),
    ];
  }
}
